src/vec.rs

```diff
@@ -120,18 +124,64 @@ named!(range_literal <CompleteByteSlice, usize>, do_parse!(
 	(num * suffix)
 ));
 
+enum VectorRange {
+	// A bare range has no subquery associated with it
+	Bare(usize),
+	Subquery(usize, Option<usize>),
+}
+
+// Parse a range that must include a subrange selector. That is, the query's range component must
+// be of the form [<range>:] and may include an optional range after the colon. For example, [10m:]
+// and [1m:5m] are allowable, but a "bare" range like [10m] is not.
+fn subquery_range_selector(input: CompleteByteSlice) -> IResult<CompleteByteSlice, VectorRange> {
+	ws!(input,
+		delimited!(char!('['),
+			do_parse!(
+				range: range_literal >>
+				tag!(":") >>
+				resolution: opt!(range_literal) >>
+				( VectorRange::Subquery(range, resolution) )
+			),
+			char!(']')
+		)
+	)
+}
+
+// Parse a range vector selector. No subqueries are allowed here, it must be of the form specified
+// at https://prometheus.io/docs/prometheus/latest/querying/basics/#range-vector-selectors
+fn range_selector(input: CompleteByteSlice) -> IResult<CompleteByteSlice, VectorRange> {
+	do_parse!(input,
+				range: ws!(delimited!(
+								char!('['),
+								range_literal,
+								char!(']')
+							)) >>
+				( VectorRange::Bare(range) )
+	)
+}
+
 pub(crate) fn vector(input: CompleteByteSlice, allow_periods: bool) -> IResult<CompleteByteSlice, Vector> {
 	ws!(
 		input,
 		do_parse!(
 			labels: call!(instant_vec, allow_periods) >>
-			range: opt!(
-				delimited!(char!('['), range_literal, char!(']'))
-			) >>
+			range: opt!(alt!(call!(range_selector) | call!(subquery_range_selector))) >>
```

Comment:
I think I know one of the reasons why you got confused while implementing parametrized range parser.

Here's a thing: this change is not enough for `label_replace(…) [5m:]` to be accepted as a valid expression since `label_replace()` is not a vector but an expression. `vector{}` alone could be parsed as a full (subquery) expression so it accepts both `[5m]` (a vector modifier) and `[5m:]` (an expression modifier that turns it into proper subquery), but the reverse is not true (`sum()` can be parsed as an expression and thus accept `[5m:]`, but it's not a vector so `[5m]` is invalid here).

So what you should do instead is introduce `expr::Node::Subquery` and modify expression parser accordingly. (Bonus points for being consistent with Prometheus at discarding expressions like `123[5m:]` or `foo[5m][5m:]`, but it's no big deal to leave that to future pull requests.)

Reply:
Thanks for your patience here. I'm pretty new to PromQL and this particular usage threw me.

I'm working off docs at prometheus.io but it's fairly basic. Is there a PromQL language spec I should review instead? I appreciate your help and I don't want to try your patience iterating on this.

Reply:
> Is there a PromQL language spec I should review instead?

I don't know any, unfortunately, it seems docs and source code is all there is. I guess I just have written enough queries in the past to have pretty strong mental model of how this thing works (or supposed to) and why some things are the way they are (e.g. some people see `!=` and `=~` as something random and ad-hoc when in reality one can just pretend that `=` is a shorthand to `==`, and then it makes total sense: `==`/`!=`, `=~`/`!~`).